sys/kern/src/arch/riscv64.rs

```diff
@@ -0,0 +1,603 @@
+// wide, we only utilise the first 32-bits of each register, setting the
+// high-order bits to 0 on startup, and restarting the low-order bits of mtime
+// back to 0 on each interrupt.
+//
+#[no_mangle]
+unsafe fn set_timer(tick_divisor: u32) {
+    // Set high-order bits of mtime to zero.  We only call this function prior
+    // to enabling interrupts so it should be safe.
+    unsafe {
+        asm!("
+        li {0}, {mtimecmp}  # load mtimecmp address
+        sd {1}, 0({0})      # set mtimecmp register
+
+        li {0}, {mtime}     # load mtime address
+        sd zero, 0({0})     # set low-order bits back to 0
+        ",
+            out(reg) _,
+            in(reg) tick_divisor,
+            mtime = const MTIME,
+            mtimecmp = const MTIMECMP,
+        );
+    }
+}
+
+#[allow(unused_variables)]
+pub fn start_first_task(tick_divisor: u32, task: &task::Task) -> ! {
+    // Configure MPP to switch us to User mode on exit from Machine
+    // mode (when we call "mret" below).
+    unsafe {
+        register::mstatus::set_mpp(MPP::User);
+    }
+
+    // Write the initial task program counter.
+    register::mepc::write(task.save().pc as *const usize as usize);
+
+    //
+    // Configure the timer
+    //
+    unsafe {
+        CLOCK_FREQ_KHZ = tick_divisor;
+
+        // Reset mtime back to 0, set mtimecmp to chosen timer
+        set_timer(tick_divisor - 1);
+
+        // Machine timer interrupt enable
+        register::mie::set_mtimer();
+
+        // Global machine interrupt enable
+        register::mstatus::set_mie();
+    }
+
+    // Load first task pointer, set its initial stack pointer, and exit out
+    // of machine mode, launching the task.
+    unsafe {
+        CURRENT_TASK_PTR = Some(NonNull::from(task));
+        asm!("
+            ld sp, ({sp})
+            mret",
+            sp = in(reg) &task.save().sp,
+            options(noreturn)
+        );
+    }
+}
+
+impl crate::atomic::AtomicExt for AtomicBool {
+    type Primitive = bool;
+
+    #[inline(always)]
+    fn swap_polyfill(
+        &self,
+        value: Self::Primitive,
+        ordering: Ordering,
+    ) -> Self::Primitive {
+        self.swap(value, ordering)
+    }
+}
+
+/// Records the address of `task` as the current user task.
+///
+/// # Safety
+///
+/// This records a pointer that aliases `task`. As long as you don't read that
+/// pointer while you have access to `task`, and as long as the `task` being
+/// stored is actually in the taask table, you'll be okay.
+pub unsafe fn set_current_task(task: &mut task::Task) {
+    // Safety: should be ok if the contract above is met
+    // TODO: make me an atomic
+    unsafe {
+        CURRENT_TASK_PTR = Some(NonNull::from(task));
+    }
+}
+
+#[used]
+static mut TICKS: u64 = 0;
+
+/// Reads the tick counter.
+pub fn now() -> Timestamp {
+    Timestamp::from(unsafe { TICKS })
+}
+
+#[allow(unused_variables)]
+pub fn disable_irq(n: u32) {}
+
+#[allow(unused_variables)]
+pub fn enable_irq(n: u32) {}
```

Comment:
May be off-topic, but is the plan to leave notifications as 32-bit for this port?

Reply:
Yes, I don't think we would need more than 32 notifications per task. So, for now leaving this at 32-bit seems fine.

Reply:
Sounds good.